src/ansi-c/library/cprover_contracts.c

```diff
@@ -0,0 +1,4116 @@
+  // object set supporting the is_freshr predicate checks (by object id)
+  __CPROVER_contracts_obj_set_t is_freshr_seen;
+  // object set recording the is_freshr allocations in post conditions
+  // (replacement mode only)
+  __CPROVER_contracts_obj_set_ptr_t linked_allocated;
+  // true iff this write set is used for contract replacement
+  __CPROVER_bool replacement;
+  // true iff this write set checks requires clauses in an assumption ctx
+  __CPROVER_bool assume_requires_ctx;
+  // true iff this write set checks requires clauses in an assertion ctx
+  __CPROVER_bool assert_requires_ctx;
+  // true iff this write set checks ensures clauses in an assumption ctx
+  __CPROVER_bool assume_ensures_ctx;
+  // true iff this write set checks ensures clauses in an assertion ctx
+  __CPROVER_bool assert_ensures_ctx;
+} __CPROVER_contracts_write_set_t;
+
+typedef __CPROVER_contracts_write_set_t *__CPROVER_contracts_write_set_ptr_t;
+#endif
+
+inline __CPROVER_contracts_car_t
+__CPROVER_contracts_car_create(void *ptr, __CPROVER_size_t size)
+{
+#pragma CPROVER check push
+#pragma CPROVER check disable "pointer"
+#pragma CPROVER check disable "pointer-primitive"
+#pragma CPROVER check disable "unsigned-overflow"
+#pragma CPROVER check disable "signed-overflow"
+#pragma CPROVER check disable "undefined-shift"
+#pragma CPROVER check disable "conversion"
+  __CPROVER_assert(
+    ((ptr == 0) | __CPROVER_rw_ok(ptr, size)),
+    "ptr NULL or writable up to size");
+  __CPROVER_contracts_car_t car = {0};
+  car.is_writable = ptr != 0;
+  car.size = size;
+  car.lb = ptr;
+  car.ub = (void *)((char *)ptr + size);
+  __CPROVER_assert(
+    __CPROVER_same_object(car.lb, car.ub), "no offset bits overflow on ub");
+#pragma CPROVER check pop
+  return car;
+}
+
+/* FUNCTION: __CPROVER_contracts_car_set_create */
+
+#ifndef __CPROVER_contracts_write_set_t_defined
+#  define __CPROVER_contracts_write_set_t_defined
+
+typedef struct __CPROVER_contracts_car_t
+{
+  __CPROVER_bool is_writable;
+  __CPROVER_size_t size;
+  void *lb;
+  void *ub;
+} __CPROVER_contracts_car_t;
+
+typedef struct __CPROVER_contracts_car_set_t
+{
+  // maximum number of elements that can be stored in the set
+  __CPROVER_size_t max_elems;
+  // 1 + greatest used index in elems,
+  // next usable index in elems if less than max_elems
+  __CPROVER_size_t watermark;
+  // A array of car_t of size max_elems
+  __CPROVER_contracts_car_t *elems;
+} __CPROVER_contracts_car_set_t;
+
+typedef __CPROVER_contracts_car_set_t *__CPROVER_contracts_car_set_ptr_t;
+
+typedef struct __CPROVER_contracts_obj_set_t
+{
+  // maximum number of elements that can be stored in the set
+  __CPROVER_size_t max_elems;
+  // 1 + greatest used index in elems,
+  // next usable index in elems if less than max_elems
+  __CPROVER_size_t watermark;
+  // number of elements currently in the elems array
+  __CPROVER_size_t nof_elems;
+  // true iff nof_elems is 0
+  __CPROVER_bool is_empty;
```

Comment:
Why store this if it is as simple as (nof_elems == 0)? It just seems like there is an invariant there that we have to be careful to preserve now.